specs/_features/eip7549/beacon-chain.md

```diff
@@ -0,0 +1,138 @@
+# EIP-7549 -- The Beacon Chain
+
+## Table of contents
+
+<!-- TOC -->
+<!-- START doctoc generated TOC please keep comment here to allow auto update -->
+<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
+
+- [Introduction](#introduction)
+- [Preset](#preset)
+- [Containers](#containers)
+  - [Modified containers](#modified-containers)
+    - [`Attestation`](#attestation)
+    - [`IndexedAttestation`](#indexedattestation)
+- [Helper functions](#helper-functions)
+  - [Misc](#misc)
+    - [`get_committee_indices`](#get_committee_indices)
+  - [Beacon state accessors](#beacon-state-accessors)
+    - [Modified `get_attesting_indices`](#modified-get_attesting_indices)
+  - [Block processing](#block-processing)
+    - [Modified `process_attestation`](#modified-process_attestation)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+<!-- /TOC -->
+
+## Introduction
+
+This is the beacon chain specification to move the attestation committee index outside of the signed message. For motivation, refer to [EIP-7549](https://eips.ethereum.org/EIPS/eip-7549).
+
+*Note:* This specification is built upon [Deneb](../../deneb/beacon_chain.md) and is under active development.
+
+## Preset
+
+| Name | Value | Description |
+| - | - | - |
+| `MAX_ATTESTER_SLASHINGS` | `2**0` (= 1) |
+| `MAX_ATTESTATIONS`       | `2**3` (= 8) |
+
+## Containers
+
+### Modified containers
+
+#### `Attestation`
+
+```python
+class Attestation(Container):
+    aggregation_bits: List[Bitlist[MAX_VALIDATORS_PER_COMMITTEE], MAX_COMMITTEES_PER_SLOT]  # [Modified in EIP7549]
```

Comment:
ie we should be able to flatten this to a single list which would save on merkle tree size / hashing time, data size, redundancy in encoding etc etc - important for lots  of reasons given the hundreds of thousands of attestations we have to process

Reply:
Having
```
aggregation_bits: Bitlist[MAX_VALIDATORS_PER_COMMITTEE * MAX_COMMITTEES_PER_SLOT]
```
will make the uncompressed size of Attestation outside of block (unaggregated + aggregated) very big. This List has exactly 1 element except when included in a blog after which it's likely to be fully populated. With your suggestion the Bitlist will be 99% zeros while on p2p topics. 

SSZ serialization involves an extra offset for the List wrapper. SSZ merkleization will have ~8 extra depth which is indeed inefficient but not sure if it's that bad.

Reply:
This begs the question: should we use a simpler wrapper for spreading attestations (vs the ones that end up in the block) on the single-attestation topics?

Reply:
As a developer I like such cleaner distinction but it has marginal benefit and does not justify adding a new container IMO

Reply:
well, single-vote attestations are some of our most cpu-and-network-demanding objects - in this case, they are trivially mappable to a gossipsub topic as well meaning they don't really "spread" beyond the gossip spec

Reply:
The optimized container only shaves a few bytes, won't have a meaningful impact on bandwidth nor CPU, you still have to verify a BLS sig

Reply:
Well apart from bandwidth and compression time, it also shaves hashing for the message id - true, not the heaviest thing, but still appears when profiling when I look at it (specially on `subscribe-all-subnets` nodes).

We can "internally" reduce the memory footprint that an extra `List` indirection adds, for the purpose of building an attestation pool, but there are these observable aspects to it which prevent it from being done across the board.

I would not suggest this if we weren't changing Attestation already.

Reply:
Also, notably, this would enforce the rule that the gossip topic should have only one signatory, simplifying the gossip validation rules.

Reply:
fwiw, `Attestation` data structure in attestation subnets keeps open optimisation path where every attesting node publishes a single aggregate instead of multiple single attestations. Although, it does require either core protocol or network layer changes but switching to `SingleAttestation` would take a step back from this path

Reply:
> keeps open optimisation path where every attesting 

this is a bridge to cross when it's time to cross it, I suspect - ie one reason we don't allow it already is because it's hard to guarantee disjoint attestation sets like that - the single attestation is fundamentally more simple to reason about and it already exists as a "concept" in the protocol, even if it doesn't have its own container.